pkg/generator/helper/helper.go

```diff
@@ -188,35 +188,130 @@ func Test{{.CobraCmdFuncName}}(t *testing.T) {
 
 const stepGoImplementationTemplate = `package cmd
 import (
+	"fmt"
 	"github.com/SAP/jenkins-library/pkg/command"
 	"github.com/SAP/jenkins-library/pkg/log"
 	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"github.com/SAP/jenkins-library/pkg/piperutils"
 )
 
+type {{.StepName}}Utils interface {
+	command.ExecRunner
+
+	FileExists(filename string) (bool, error)
+
+	// Add more methods here, or embed additional interfaces, for everything you need to be able to mock in tests.
+	// Unit tests shall be executable in parallel (not depend on global state), and don't (re-)test dependencies.
```

Comment:
Sorry, but I don't follow you. There is a clear benefit of executing tests in parallel, which is shorter overall test runtime. At the point **when** we would have a clear problem with test execution time, it would be quite a big effort to refactor all code to not depend on global state, don't you think? Why not do it the proper way from the start?

Reply:
To elaborate a bit more on global state: You probably agree that tests should not depend on execution order, if nothing else then for the simple reason that you should be able to run a test in isolation. This could be rephrased as "tests should not depend on each other". This extends to "tests should not depend on the correctness of other tests". To put that another way: Making tests depend on global state is just one more way to write tests incorrectly and influence/break other tests. As an additional draw-back, they cannot be executed in parallel, increasing runtime. So while your statement that there is nothing wrong with executing tests sequentially, it seems to support the idea that there is nothing wrong with relying on global state in tests. Hopefully I provided a clear line of thought on why I don't agree with that.

Reply:
I've found https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html a very interesting read.

Reply:
Of course tests should be fully independent, a test must never depend on the result from another test. From that point of view test should be able to run in parallel.

But saying tests are strongly independent from each other and therefore they _must_ run in parallel is another thing. There might be other reasons beside independent tests why tests do not run in parallel. There might be other valid reasons for having global vars. Your doctrine of avoiding global vars totally is IMO false. We should keep the freedom of having global vars. I know that tests influences the style of the code under tests. But requiring the absense of global vars for the sake of being able to run test in parallel is IMO to much. 

I think there is nothing wrong with having some tools like shell handlers, http handlers as global vars. This ensures smaller signatures and helps IMO  having clearer signatures resulting in slighty better understandable and maintainable code. That can be improved by collecting all the handlers into a bundle so that only the bundle needs to be forwarded through all the methods. An in turn this results in a more complex bundle type.

When having global vars we need to ensure that any state kept inside the corresponding mocks is cleared before the next test is launched. But that is possible, I don't see a bigger issue here.

With regards to the link shared above: In the post it is suggested to forward even the logger through the signature

```
func NewObject(q RowQueryer, r Requestor, n int, logger log.Logger)
```

I doubt this is an improvement.

I know a lot of cases where someone pushes some ideas in partly very exalted ways. I think keeping a more moderate and reasonable habit it in most cases better. The pendulum swings to the extrem points, but maybe a more balanced way is in the long run the better way.

Reply:
I agree with you about avoiding more complicated signatures. But combining all required dependencies into one interface does take care of that. You could argue that the logger is as "hidden" as before from the signature, because it does not become an additional parameter. However, the difference is that it is explicit via the interface, and it can be mocked. On numerous occasions, I wanted to test log output of error paths (because that is essentially our user interface), improving coverage, but often decided against it because of the considerable overhead when setting up tests, and also because I would then ruin parallel execution. So I do not agree with you that a line can be drawn and logging never needs to be mocked. This is really no theoretical discussion for me, I found it cumbersome that we access the logger directly many times in practice.

Reply:
> and logging never needs to be mocked

I never said that log output should not asserted. I know some cases where log output was asserted instead of some real action. That is actually bad (eg assert a log message next to e.g. some shell call where the log message was asserted instead of the shell call as some nearby hint), but I also see valid reason to assert important log messages.

But this does not mean that a logger should be contained in all the signatures. That is basically boring. It is possible to mock global vars. Yes, in that case parallel test execution is an issue, but I do not fully understand why parallel test executing gets that important out of the middle of nowhere. I don't say that parallel test execution is bad, but there are other dimensions beside that (e.g. clearer signatures).

> But combining all required dependencies into one interface does take care of that

That ends up in having a complex util used for shell access, managing http traffic,  file system access and logging. That is also not the yellow of the egg. Now you can start with introducing a better structure for that util. But is that really better compared with having global vars?

Reply:
> But saying tests are strongly independent from each other and therefore they _must_ run in parallel is another thing. There might be other reasons beside independent tests why tests do not run in parallel. There might be other valid reasons for having global vars. Your doctrine of avoiding global vars totally is IMO false. We should keep the freedom of having global vars. I know that tests influences the style of the code under tests. But requiring the absense of global vars for the sake of being able to run test in parallel is IMO to much.

If another coding pattern is just as easy to work with, while avoiding any global state and thereby giving two clear benefits, why not suggest it as a starting point for new steps? The comment says "Unit tests **shall** be ...", but "must be", so it is a strong suggestion, not a strict requirement.

> I think there is nothing wrong with having some tools like shell handlers, http handlers as global vars.

But this contradicts accumulated wisdom, not only from the article I linked.

> This ensures smaller signatures and helps IMO having clearer signatures resulting in slighty better understandable and maintainable code. That can be improved by collecting all the handlers into a bundle so that only the bundle needs to be forwarded through all the methods. An in turn this results in a more complex bundle type.

I fully agree about the small signatures. Which is absolutely one of the goals of this PR: providing a way to keep signatures small while making dependencies explicit and replaceable at the same time.

Reply:
> > But combining all required dependencies into one interface does take care of that
> 
> That ends up in having a complex util used for shell access, managing http traffic, file system access and logging. That is also not the yellow of the egg. Now you can start with introducing a better structure for that util. But is that really better compared with having global vars?

Can you elaborate on what you find complex about it? How is it different from having a bunch of imports and accessing functionality prepended by package name?

I think one thing we should start doing, and that is also a learning from the articles I've read, is that we should have much smaller interfaces. So by embedding only the required interfaces, you get a better picture of what functionality the step really needs. You can also embed method signatures, which would make it 100% explicit, but then you have the drawback of having to adjust more code, should a signature ever need to change. In the proposed step skeleton, I make an example of both: embedding an interface, and a method signature.

As for the resulting runtime struct implementing the interface, I don't really see what is complicated about it.

Reply:
> But this contradicts accumulated wisdom

arguing with "accumulated wisdom" is dangerous. We life in a plural world where typically several doctrins compete with each other. What is "wisdom" for the one doctrin might be kind of less important for the other doctrin. There are only rare cases when one doctine dominates anohter doctine in the sense that is it better in all dimensions.

> you get a better picture 

We can also get a good picture when we have global vars. Don't say that it is not possible to get a good picture also with a util. When we keep some practices in mind (short method bodies, small signatures, no monster inferfaces) this is not an issue also when having global vars.

> [util] what you find complex about it

* We have a util without clear scope. The util can handle file access, http traffic, can make shell calls. Now we can start to better structure the util. What is really better compared to globals? Now we are again at the point with the parallel test execution. I fully agree that tests must be independent. But that is another story. In case we have tests which cannot run in parallel is that such a big thing?

* The util needs to be forward to each and every method. I personally think it is easier to get an overview when signatures contains thinks like IDs processed in the method body, but not environment related things like file system handlers, shell handlers and so on. There is nothing wrong with having these entities as globals. In other languages we have e.g. global vars which gets the values injected by some dependency injection framework ...

For me using the utils or using global vars does not make such a big difference. Hence in my personal opinion there is not a sufficient benefit for introducing this as part of the generated code.

As you might noticed I didn't use "require changes" so simply go ahead with the approach in case I cannot convince the other parties. Don't think that we will come to a closer common understanding. And that is also not required.

Reply:
> > But this contradicts accumulated wisdom
> 
> arguing with "accumulated wisdom" is dangerous.

I agree, but let me rephrase: I googled for Best Practices with the goal of getting a broad perspective. I found a bunch of articles, some directly on golang.org. Virtually all of them argued to avoid global state and to make dependencies explicit, none argued that using globals is fine. Why should I not conclude that this is the wisdom to follow?

> We life in a plural world where typically several doctrins compete with each other. What is "wisdom" for the one doctrin might be kind of less important for the other doctrin. There are only rare cases when one doctine dominates anohter doctine in the sense that is it better in all dimensions.

Of course, when talking in these broad terms. But we specifically discuss best practices for Go code. If the above is also true for that specific topic, there must be alternative articles arguing that using package global function pointers is a perfectly good solution, at least for certain situations. I'd be very interested if you have some links.
 
>  When we keep some practices in mind (short method bodies, small signatures, no monster inferfaces) this is not an issue also when having global vars.

Even a good IDE such as GoLand will be at a loss with global function pointers in some situations.

> > [util] what you find complex about it
> 
> We have a util without clear scope.

The clear scope is:
* Make important dependencies which influence runtime behavior explicit.
* Reduce the number of required function arguments.

> The util can handle file access, http traffic, can make shell calls.

Yes, but the important question is: Does it make the code using the utils object any more complex? I argue that it doesn't. And it helps to reduce function parameter count.

> Now we can start to better structure the util. What is really better compared to globals? Now we are again at the point with the parallel test execution. I fully agree that tests must be independent.

You answer your own question.

> But that is another story. In case we have tests which cannot run in parallel is that such a big thing?

Yes.

> * The util needs to be forward to each and every method. I personally think it is easier to get an overview when signatures contains thinks like IDs processed in the method body, but not environment related things like file system handlers, shell handlers and so on.

We both agree that reducing function arguments is a good thing. The utils help with that. Until now, we the `command.Command` argument in almost every function. That one would get replaced. So the utils don't make anything worse.

> There is nothing wrong with having these entities as globals.

No, there are problems with that which were already pointed out in this discussion.

> In other languages we have e.g. global vars which gets the values injected by some dependency injection framework ...

But we are talking about Go.

> For me using the utils or using global vars does not make such a big difference. Hence in my personal opinion there is not a sufficient benefit for introducing this as part of the generated code.

Ok.

> As you might noticed I didn't use "require changes" so simply go ahead with the approach in case I cannot convince the other parties. Don't think that we will come to a closer common understanding. And that is also not required.

Ok. Thanks for your input! I hope it's Ok that I provided feedback to your points one more time. I probably don't have anything more to add, but if you can point to some articles advertising global vars as a good solution, I'd be interested. I want to keep an open mind.